cataclysmbnteam/Cataclysm-BN
# feat: Effects adding other effects on removal

## Purpose of change



This PR allows defining comedowns and similar effects that can coexist with the parent effect and can be defined by mods.

## Describe the solution



When effect expires, it can create other effects, at set duration/intensity, or inheriting duration/intensity from the decaying effect.
New effects are added at the end of the effect processing loop, and so should not be able to cause effect map migration. They are added in the same order as the parent effects were processed.
There is an option to only affect decaying (0 duration) effects, only affect removed effects (`remove_effect` when duration > 0), or both.

Reimplemented adrenaline and its comedown in the new system. This means that both effects can now coexist - you can have adrenaline buff with comedown debuff applied at the same time.

## Describe alternatives you've considered



* Waiting for lua to cover it, with associated problems, such as lack of common standard of effect decay
* Leaving it hardcoded, since we don't have many effects that could use this feature (possibly because we didn't have this feature)

## Testing



* Tests pass
* Spawn 3x adrenaline
* Use adrenaline
* Wait for it to decay
* Apply adrenaline
* Confirm both buff and debuff exist
* Wait for both to time out
* Apply adrenaline
* Use blood filtering bionic
* Confirm that adrenaline comedown is present

## Additional context



## Checklist



* [x] Body part inheritance
* [x] Fixing adrenaline:
  * [x] Rebalance duration, to account for debuff no longer being included in "common duration"
  * [x] Remove hardcoded addition of debuff from blood filtering and possibly others
